### DISCUSSION TITLE
recombination frequency

### DISCUSSION THREAD

**Discussion by Anilkumargpb** [gaynorr/AlphaSimR#187] [General]
Hi 
I want to double the recombination rate in the simulation. I am using the following code:

library(AlphasimR)
founderPop=importInbredGeno(geno = snp, genMap = GenMap, ped = NULL)
SP = SimParam$new(founderPop)
SP$setTrackRec(TRUE)
SP$setRecombRatio(2)
SP$importTrait(markerNames = qtleffects$marker,
               addEff = qtleffects$aditiveEffect,
               name = "Your_Trait")
SP$addSnpChip(10)
Parents = newPop(founderPop)
crossPlan<-matrix(c(1,2,3,4,5,2,3,4,5,1), nrow = 5, ncol = 2)
F1 = makeCross(Parents, crossPlan,  simParam = SP)
DH=makeDH(F1, 150)
Geno_cycle1<-pullSegSiteGeno(DH)

The crossing plan is a kind of chain crossing. Please let me know, the code used is correct? how do I cross check? 

Thanks 
Anil

**Comment by gregorgorjanc** [comment]
@Anilkumargpb note that  `SimParam$setRecombRatio()` sets `Set the relative recombination rates between males and females. This allows for sex-specific recombination rates, under the assumption of equivalent recombination landscapes.` (copied from the help page). If you want to double recombination *overall* rate, you have to work with your `GenMap` - see also `SimParam$switchGenMap()`.

**Comment by Anilkumargpb** [comment]
Hi Gregor,

Thanks for your response. I tried to change the recomination rate by modifying the genMap lengths. But, I could not verify the change in recombinations in the simulated populations. I really appreciate any thoughts on this. 

Thanks 
Anil

**Comment by gregorgorjanc** [comment]
Maybve count the number of recombination events? See for example https://github.com/gaynorr/AlphaSimR/issues/34

**Comment by gaynorr** [comment]
@Anilkumargpb, here is an example script showing how you can double the length of the genetic map in an AlphaSimR simulation. Doubling the genetic map distance will double the expected number of recombinations. Note that this won't necessary double the recombination rate between loci, because the relationship between genetic distance and recombination rate is not linear. The genetic map distance between loci can be freely changed at any time during the simulation, but the physical ordering of the loci must remain the same.

```
library(AlphaSimR)

# Create genotypes for two inbred parents
geno = rbind(c(2,2),
             c(0,0))
colnames(geno) = c("A", "B")

# Create genetic map for two loci 10 cM apart
genMap = data.frame(
  markerName = c("A", "B"),
  chromosome = c(1, 1),
  position = c(0, 0.1) # In Morgans
)

# Create an initial population of founder haplotypes
founderPop = importInbredGeno(geno = geno, 
                              genMap = genMap)

# Set simulation parameters
SP = SimParam$new(founderPop)

# Create parents from founder population
Parents = newPop(founderPop)

# Create 100,000 F2 progeny
F1 = makeCross(Parents, crossPlan=cbind(1,2))
F2_10cM = self(F1, nProgeny=100000)

# Measure observed recombination frequency
H_10cM = pullSegSiteHaplo(F2_10cM)
sum(H_10cM[,1] != H_10cM[,2]) / nrow(H_10cM)

# Expected recombination rate from the Kosambi mapping function
# AlphaSimR should roughly approximate this rate with its default gamma model interference parameter set to v=2.6
# Note you could set "SP$v=1" to switch to no crossover interference and effectively a Haldane mapping function
# AlphaSimR doesn't convert map distance automatically when changing the parameter, so this may need to be done 
# manually.
0.5 * (exp(4*0.1)-1) / (exp(4*0.1)+1)


# Double the spacing of marker on genetic map
# This doubles the expected number of crossover (not recombination rate)
genMap = SP$genMap
genMap = lapply(genMap, function(x) 2*x)
SP$switchGenMap(genMap=genMap)

# Create new F2 population with 20 cM spacing
F2_20cM = self(F1, nProgeny=100000)

# Measure observed recombination frequency
H_20cM = pullSegSiteHaplo(F2_20cM)
sum(H_20cM[,1] != H_20cM[,2]) / nrow(H_20cM)

# Expected recombination rate from the Kosambi mapping function
0.5 * (exp(4*0.2)-1) / (exp(4*0.2)+1)
```

**Reply by Anilkumargpb** [reply]
Hi @gaynorr 

This is great!!!
This is what I was looking for. Thank you so much. 

Anil

**Reply by Anilkumargpb** [reply]
Hi @gaynorr 

I tried to get the number of recombinaitons following your codes (provided below). To my surprise, I dont see any recombinations. I am trying this in maize which is having ten chromosomes. I am using 1200+ seg sites. I tried by increasing map length in 2, 4, 6 and 10 folds. To my surprise, I dont see any recombinations in any fold!!!. 


founderPop=importInbredGeno(geno = snp, genMap = GenMap, ped = NULL)
SP = SimParam$new(founderPop)
SP$setTrackRec(TRUE)
SP$importTrait(markerNames = qtleffects$marker,
               addEff = qtleffects$aditiveEffect,
               name = "Your_Trait")
SP$addSnpChip(10)
Parents = newPop(founderPop)
crossPlan<-matrix(c(1,2), nrow = 1, ncol = 2)
F1 = makeCross(Parents, crossPlan,  simParam = SP)
F1_rec = pullSegSiteHaplo(F1)
sum(F1_rec[,1] != F1_rec[,2]) / nrow(F1_rec)

genMap = SP$genMap
genMap = lapply(genMap, function(x) 10*x)
SP$switchGenMap(genMap=genMap)
F2=self(F1, 1500)
Geno_cycle1<-pullSegSiteGeno(F2)

F2_rec = pullSegSiteHaplo(F2)
sum(F2_rec[,1] != F2_rec[,2]) / nrow(F2_rec)

All its showing is zero recombinations. I really appriciate any suggestions from you. 

Thanks 
Anil

**Reply by gregorgorjanc** [reply]
@Anilkumargpb you are comparing alleles at the 1st and 2nd locus, which might differ or not in your data compared to @gaynorr example where he had `c(2, 2)` and `c(0,0)`. You can also look at using `pullIbdHaplo()` (parses `recHist` output I mentioned above) and look at further distance between loci - you will have limited number of recombinations between two neighbouring loci.

**Reply by Anilkumargpb** [reply]
Hi @gregorgorjanc,

I am confused about finding the total recombinations or I am missing something. I am trying to find total number of genome wide recombinations in a simulated maize population. Please suggest me the better way of doing it. 

Thanks 
Anil

**Reply by gregorgorjanc** [reply]
Notice how many progeny (and hence opportunities for recombination) Chris has created and how many you have created.

Do you have polymorphism at the pairs of loci that you are looking at? You are looking at recombination between 1st and 2nd locus after Chris’ example where he started with opposite genotypes for two individuals. Does your `snp` genotype object have variation at the first two loci? If you don’t, then you won’t see recombination even if one happens since all alleles are the same.

You will likely want to look across the genome and not just first two loci too.

**Reply by Anilkumargpb** [reply]
Thanks for you time!!

Yes, I obeserved that my first two loci are not polymorphic. However, most of segsites are polymorphic between parents. As you mentioned, my objective is to look across the genome but I could not do that. I tried the example at #34, but its I did not see total recombinations. Using recHist is more appropriate for achieving my objective? Please suggest me.

Thanks 
Anil

**Reply by gregorgorjanc** [reply]
@Anilkumargpb the post at https://github.com/gaynorr/AlphaSimR/issues/34 describes how AlphaSimR stores recombination events. You could parse that `recHist` object and count how many recombinations happened to get the "total" number.

**Reply by Anilkumargpb** [reply]
I follow the recHist object to count the recombination across genome. 

Thank you so much 
Anil
